### DISCUSSION TITLE
flow typing for eliminated parts of union types in default match case?

### DISCUSSION THREAD

**Discussion by Swoorup** [scala/scala3#17844] [Feature Requests]
In typescript I can do like the following: 

```typescript
type U = "A" | "B" | "C"
function uBut(x: U): "B" | "C" {
  switch (x) {
    case "A":
      return "B"
    default:
      return x
  }
}
```

Here, `x` is deduced to a union of `"B" | "C"`. 

With scala 3 union unfortunately you can't do so at the moment. The following fails to compile:

```scala
type U = Int | String | Boolean

def uButString(u: U): Int | Boolean =
  u match
    case s: String => 0
    case x => x
```

This will lead people to instead use `asInstanceOf` which is bad imo. You can still make it typesafe if you annotate the last type, without warnings like

```scala
type U = Int | String | Boolean

def uButString(u: U): Int | Boolean =
  u match
    case s: String => 0
    case x: (Int | Boolean) => x
```

But it would be nice, if scala was able to auto deduce the type of `x` to `Int | Boolean`, leading to more natural code and less usage of `asInstanceOf`.

If there was a condition on the pattern match, we could retain the current behavior as is.

**Comment by dwijnand** [comment]
As a side note: the difference between the two is that the default case makes it also match and return `null`, which with `MatchError` in the second example.
